### DISCUSSION TITLE
9.10a3 General Purpose Block Feedback

### DISCUSSION THREAD

**Discussion by facelessuser** [facelessuser/pymdown-extensions#1940] [Show and tell]
# Alpha 3
Documentation: https://facelessuser.github.io/pymdown-extensions/extensions/blocks/

The purpose of this discussion is to allow feedback for the latest iteration of General Purpose Blocks 9.10.a3. It should be noted that any and all feedback is welcome, but the current focus is on the API for creating block plugins.

## Changes in Alpha 3:

- **NEW**: General blocks now use an indented option block right after the header. `yaml_indent` option has been
  removed.
- **NEW**: Added new "Definition" block that allows the creation of definition lists.
- **NEW**: Simplified argument configuration.
- **NEW**: Some internal cleanup.
- **NEW**: Documented current API.

### Syntax

We still have the following open syntax question. Currently, we have one general option that is only available during the alpha (and maybe beta) stage. By the final release, we'd like to remove this option with a permanent solution.

1. Which is preferred, syntax with `:::` or `///`? Currently, the option `colon_syntax` allows selecting between the current default syntax of `///` and the colon syntax (`:::`). Which is generally preferred by the community?

### API

We are looking for general feedback on the API.

1. Does this look sufficient or are we missing some cases that adjustments to the API could help better handle?
2. Are there any confusing aspects of the API? Do we need more clarification or even more simplification of certain parts of the API?
3. etc.

**Comment by waylan** [comment]
> * Added new "Definition" block that allows the creation of definition lists.

Didn't see this in #1883 until today. So I'm following up here. I find this a weird one because there is an established (and widely used across implementations) definition list syntax. Of course, it isn't supported everywhere and I know some don't care for it. And anyone is certainly free to define their own syntax with their on extension. I just find it odd to include this in the default set of supported blocks. I would expect this to be separate. But maybe that is just me. I have nothing more to say about it. Of course, this is your project so you can do what you want.

**Reply by facelessuser** [reply]
This is new, and I am aware that some parsers respect definition lists as Python Markdown defines them, but there are certainly many that do not.  I know in my editor it is constantly thinking that additional paragraphs under a definition are code blocks. It is kind of annoying because they aren't widespread enough.

Indeed, no one is required to opt into all the blocks that will be provided. We can certainly discuss defaults as well. In this case, I imagine that what we've implemented can coexist with definition lists in the sense that if you do not use them both (physically use both syntaxes) that having them both enabled should cause no trouble.

The only conflict you could possibly have is if you are physically defining general block definition lists, but then also have the original "definition lists" enabled and follow up with an indented block. In this case, I could see the original definition lists trying to append it to the description of the `<dl>` element right before it, but then I'd ask, why are you defining general block definitions but also enabling the alternative form of definition lists. In this case, I'd recommend excluding the original "definition lists".

```
/// define
term

- description
///

    some block
```

Again, one of the reasons I have an interest in this plugin is to eliminate non-standard indentation syntax from my own documents but is not something I am forcing on others.

**Reply by facelessuser** [reply]
It may make sense to maybe make most or all opt-in. Maybe only include the generic HTML block by default and all others are added to it...

**Reply by waylan** [reply]
> It may make sense to maybe make most or all opt-in. Maybe only include the generic HTML block by default and all others are added to it...

This may be a good way to start. Then, over time with more feedback, it may be clear that users are always using some block types more regularly. If so, they could be included by default in a future release.

**Comment by waylan** [comment]
The API looks pretty comprehensive. After an initial read through the documentation, here are my observations.

I'm struggling to see the difference between the `on_init` and `on_parse` events. Obviously, `on_parse` can return `False` which causes the block to fail (btw, what happens when it fails?). Other than that, they seem to be identical. My best guess is that perhaps the args and options available in `on_init` are the default options for the extension whereas the args and options available in `on_parse` include the args and options pulled from the block itself. If so, that was not clear to me from the documentation. I am only guessing that because the name `on_init` suggests parsing hasn't happened yet which means the args and options haven't been retrieved from the block.  If that is not the case, then I would suggest consolidating the two into `on_parse`.

The `on_markdown` event also confused me at first. Then I realized it was returning a string (`-> str:`). Perhaps the first column of the table should be labeled `return values` (or similar) rather than `options`. Options sound like something that should be passing in, not returned. And perhaps more explicit text in the description would help here as well.

A minor issue, but the `type_string` validator is defined twice in the documentation. Once before and once after `type_insensitive_string`.

I find it curious that the `type_html_attribute_value` (and I assume the other validators which act on HTML attributes) escape the values as that happens by the [serializer](1) after the etree is complete. I realize that the serializer is smart enough to avoid escaping existing entities (`&gt;` does not become `&amp;gt;`), so this should not break things, but why not let the serializer take care of this?

Finally, wow that is a lot of validators. At first I was going to object as YAML already has types, but now I see you are doing a lot more than differentiating between strings, integers, and bool values. This is extremely powerful. I'm not certain all of them will be widely used, but they exist now so no reason to remove them. And of course, as users can define their own, you are not limiting users to the default set. Two things I don't see in the documentation are any direction on how to create your own validator and the import path of the existing validators.

Overall I think this looks like a good API. However, I haven't looked at the code or attempted to use it yet. So my impressions could change. But I'm not sure when I'll have time for that.

[1]: https://github.com/Python-Markdown/markdown/blob/master/markdown/serializers.py#L99-L104

**Reply by facelessuser** [reply]
> I'm struggling to see the difference between the `on_init` and `on_parse` events. Obviously, `on_parse` can return `False` which causes the block to fail (btw, what happens when it fails?). Other than that, they seem to be identical. My best guess is that perhaps the args and options available in `on_init` are the default options for the extension whereas the args and options available in `on_parse` include the args and options pulled from the block itself. If so, that was not clear to me from the documentation. I am only guessing that because the name `on_init` suggests parsing hasn't happened yet which means the args and options haven't been retrieved from the block. If that is not the case, then I would suggest consolidating the two into `on_parse`.

This is a good point, and maybe these can be combined as a single `on_init` or `on_parse` event. I'll take a look.

> The `on_markdown` event also confused me at first. Then I realized it was returning a string (`-> str:`). Perhaps the first column of the table should be labeled `return values` (or similar) rather than `options`. Options sound like something that should be passing in, not returned. And perhaps more explicit text in the description would help here as well.

I'll admit I didn't know exactly what to call this event. I'm open to suggestions. I think the suggestion regarding the table sound good.

> A minor issue, but the `type_string` validator is defined twice in the documentation. Once before and once after `type_insensitive_string`.

Noted, thanks!

> I find it curious that the `type_html_attribute_value` (and I assume the other validators which act on HTML attributes) escape the values as that happens by the [serializer](1) after the etree is complete. I realize that the serializer is smart enough to avoid escaping existing entities (`&gt;` does not become `&amp;gt;`), so this should not break things, but why not let the serializer take care of this?

The plugin will use these values to manually create the elements via etree, if this is already handled and I can safely remove all transforms, then I'd be happy to simplify this. I will look into this closer. This may have been oversight on my part. There may be times when  a plugin may want to know exactly what the value will be, so I'll think about this.

> Finally, wow that is a lot of validators. At first I was going to object as YAML already has types, but now I see you are doing a lot more than differentiating between strings, integers, and bool values. This is extremely powerful. I'm not certain all of them will be widely used, but they exist now so no reason to remove them. And of course, as users can define their own, you are not limiting users to the default set. Two things I don't see in the documentation are any direction on how to create your own validator and the import path of the existing validators.

People could certainly do all the validation in `on_parse` manually, the idea was to make it easy enough to use by just dropping in existing validators and provide most generic types that an author might need. This is not exhaustive and most people may not need some of these. We could certainly drop the ideas of validators altogether and just let people validate themselves in `on_parse`, but I figured this approach may be more attractive. I'm open to discuss this further if people want to.

This is actually really good feedback, so I'll take a look this week and see about cleaning things up.

**Reply by facelessuser** [reply]
I will note that `on_init` is performed at the end of the normal `__init__()` and the idea is that it could be utilized to handle global options for the plugin.  `on_parse` occurs at the time per block argument and option parsing. So there is a distinct difference. Maybe I can make this real clear. Could these technically be combined? I don't think I see a reason they couldn't be, but I may leave this separation and clarify the difference more in the documentation, and maybe provide suggestions in what kind of things could be done in the events.

**Reply by facelessuser** [reply]
I also realized I did not answer one of the questions. If `on_parse` returns `False`, the current block under evaluation will not be parsed as a generic block. It will then be parsed by other Markdown plugins. This is a way to completely invalidate the block.

**Reply by waylan** [reply]
Ah, so `on_init` runs once per document, but `on_parse` runs once per block. Is that it?

**Reply by facelessuser** [reply]
Sort of, but not quite. We kind of need an instance per block we run into. Since these can be nested, you could have multiple blocks in different states. So, we instantiate a new block every time we encounter a generic block. `on_init` is meant for global options and initializing the block to some sane default state. `on_parse` is where the per block options and arguments are gathered allowing you to validate them and make decisions based on them, potentially updating class variables and such to enable/disable features.

**Reply by facelessuser** [reply]
I've cleaned up documentation related to `on_init`, `on_parse`, and `on_markdown`. No functional changes have been made at this time.

**Reply by waylan** [reply]
That clears things up. Now I understand. I would suggest putting a polished version of that explanation in the documentation somewhere. It did not occur to me that each block get's its own instance of the class. Although, in retrospect, some of the documented behaviors only make sense with that behavior.

**Reply by facelessuser** [reply]
Done

**Comment by facelessuser** [comment]
I'm thinking about changing the argument delimiter from `|` to just `/`. Anyone have strong objections or other suggestions?

The only reason I'm considering this change is that some syntax highlighters can think that the declaration is the start of a table and can affect some highlighting afterwards. I haven't found any actual Markdown parsers that get confused by this, but I'd like the final syntax to be as un-intrusive as possible.

So, as a summary, the change would be changing this:

```
/// block | argument
```

to 

```
/// block / argument
```
Again, I'm open for people to suggesting something different, but the main options I see are encapsulating the block name with something (quotes, brackets, etc.) or using some delimiter. I think I'm okay with just changing `|` to `/`.

**Reply by waylan** [reply]
Sigh! I really dislike this change. The conflict with highlighters is unfortunate. It would be preferable to design a syntax without consideration to outside forces like this. But we don't get to work in our idealized world.  Unfortunately, I don't have any suggestions for a workaround.

**Reply by facelessuser** [reply]
It's just something I'm considering. I'll think about it. My preference was always `|`. And the annoyance in my editor is very minor. I posed this as a question as I suspect that there may be some dislike.

**Reply by facelessuser** [reply]
I played with this more. It's an odd case in my editor, that requires a very specific case as most cases it does not trigger. So I'm going to not make this change, especially for such a minor cosmetic change that, even in my case, is very finicky to reproduce.

**Comment by facelessuser** [comment]
The next version will be Beta 1.  We've cleaned up some API things and documented some missing parts. Assuming no issues, I plan to release generic blocks to the public in an official release, but marked as "experimental".

Some changes coming:

1. We are removing `colon_syntax` option and moving forward with the `///` syntax.

2. We've revised the HTML block to accept Emmet style syntax for specifying tags with attributes.I updated my documentation using the new general blocks in most places. I found it made specifying tags less of a chore when using the block. It made sense to streamline it for a block that is focused on creating HTML blocks.

    ```
    /// html | div#id.class[name=value]
    content
    ///
    ```

3. Specifying attributes on other non-HTML specific blocks was a bit laborious as well, so we shortened the `attributes` to just `$` as it will likely be common to use, and less typing is preferred, at least to me. Also, `$` takes attributes the same as HTML blocks.

    ```
     /// admonition | Title
        $: .warning#id
    content
    ///
    ```

4. We also dropped the `type` option on Admonition and Details as it was kind of redundant as you can just use `$` to set classes.

Feel free to get in your last remarks, opinions criticisms. I hope to create the beta either today or tomorrow. Final input will be considered for the official release.

**Reply by waylan** [reply]
👍 for item 1.

As for the rest, I dislike them all. Although, if I am understanding correctly, they are optional shortcuts and I can still use the traditional key/value pairs as before right?

Btw, I do not see any of these changes included in the published documentation. Maybe if I saw some clear documentation I would feel differently about them.

Finally, it had been a while since I have looked at the documentation for the specific block types and I found the config settings laborious. Take this for example:

```python
import markdown
from pymdownx.blocks.admonition import Admonition

md = markdown.Markdown(
    extensions=['pymdownx.blocks'],
    extension_configs={
        'pymdownx.blocks': {
            'blocks': [Admonition],
            'block_configs': {
                "admonition": {
                    'types': ['note', 'warning', 'some-custom-type']
                }
            }
        }
    }
)
```

I will note that my preferred way to pass config options to extensions is to pass in a class instance rather than a string name. So I would do this in the documentation:

```python
import markdown
from pymdownx.blocks import Blocks
from pymdownx.blocks.admonition import Admonition

md = markdown.Markdown(
    extensions=[Blocks(blocks=[Admonition(types=['note', 'warning', 'some-custom-type'])])],
)
```
That feels a lot less intimidating to me (perhaps some line breaks could make it easier to read). Although, I haven't actually looked at the code so I'm assuming that works. If not I would be very disappointed.

**Reply by facelessuser** [reply]
As far as HTML blocks are concerned, I probably won't be budging on this as using this in the real world, I find this far easier to use:

```
/// html | div.class#id
content
///
```

I may be open to suggestions to tweak the syntax if there are some suggestions. Though every block allows manual attribute assignment via the YAML options.

---

With that said, I may be open to the YAML option to key-value pairs via the options approach. I will likely keep that option name as `$` though. It is a reserved value, and doing so keeps it from conflicting with user-generated option names, and it is far less typing for something very commonly used.

```
/// admonition | Title
    $: {class: note}
```

I still think something like this is infinitely easier and quicker to type.

```
/// admonition | Title
    $: .note
```

I'd be interested in hearing more feedback on this point. As someone that just converted their entire document to use such things, I was not a fan of the laborious typing that the original required. I would be open to hearing some suggested tweaks or more direct criticism to the approach as well to understand what exactly why this approach is hated. Unfortunately, "I don't like it" doesn't help me understand why. Personal opinion is fickle. I've at least described why I prefer such an approach, I'd appreciate a counter argument.

---

> Finally, it had been a while since I have looked at the documentation for the specific block types and I found the config settings laborious.

Since these classes are instantiated as one for every block, it doesn't quite make sense to pass them this way. The alternative is to instantiate once and then keep a stack of states in each one, including any state specific variables. It is doable, but a little more complicated. Also, I don't think we are considering a common use case such as MkDocs. Passing objects is far more complicated. I'm now allowing string passing for extensions:

```
  - pymdownx.blocks:
      blocks:
      - pymdownx.blocks.admonition:Admonition
      block_configs:
        admonition:
          types:
          - new
          - settings
          - note
          - abstract
          - info
          - tip
          - success
          - question
          - warning
          - failure
          - danger
          - bug
          - example
          - quote
 ```

We could make Blocks take the options directly, but maybe make the `blocks` option a dictionary?

```python
md = markdown.Markdown(
    extensions=[Blocks(blocks={Admonition: {'types': ['note', 'warning', 'some-custom-type']}})]
)
```

or this if you are using strings notation:

```python
md = markdown.Markdown(
    extensions=[Blocks(blocks={'pymdownx.blocks.admonition:Admonition': {'types': ['note', 'warning', 'some-custom-type']}})]
)
```

I guess in Mkdocs it could look like this:

```yaml
  - pymdownx.blocks:
      blocks:
        pymdownx.blocks.admonition:Admonition
          types:
          - new
          - settings
          - note
          - abstract
          - info
          - tip
          - success
          - question
          - warning
          - failure
          - danger
          - bug
          - example
          - quote
 ```

**Reply by waylan** [reply]
> Unfortunately, "I don't like it" doesn't help me understand why. Personal opinion is fickle. I've at least described why I prefer such an approach, I'd appreciate a counter argument.

Fair enough. 

Let's focus on the `$` syntax first. Assume I am a new user who has never encountered this syntax and I am attempting to make a contribution to your open source project. I open your Markdown file and see this. What is this mystery `$` key? Sure, from the value I can make some assumptions, but I am only guessing. Strangely, on this other type of block the same key also appears but the value assigned to it seems to suggest some other meaning than I assumed for the first one. I am likely to give up and not make my contribution.

And why specifically `$`? As used elsewhere `.` indicates `class`, not `$`. Of course, using `.` alone as a key would look weird, but I thought the same when I saw `$` with the added problem that it is meaningless to me.

In my opinion `$` is just too abstract. `class` (and/or `type`) is so much more clear and it is only four more characters of typing. And with a good text editor with decent auto-completion, you should only have to type out the entire word once per document.

To be clear, this is enough of an issue for me that I would be inclined to write my own custom block which implements admonitions using the previous syntax (the `class` keyword). That is , unless you supported both keywords (`$` and `class`), in which case I would always only ever use `class`.

Regarding removing the `type` keyword from admonitions, I understand the motivation. However, I wouldn't remove the keyword. Continue to support it but make it unnecessary to use except when a custom type is not defined in the config. Some times there could be a one-off that doesn't justify editing the global config. A simple `type` defined in the document itself would suffice. I don't know, maybe you are already doing this, but it was not clear to me from the explanation provided.

Regarding the `div.class#id` syntax, that is purely my opinion. My opinion has always been that the option on the first line (after the `|`) should only ever be a single word. While the syntax does result in no whitespace (which is technically a single word from one point of view) it is much more complex that anything I think belongs there. This is the same reasoning I used previously to push back against you (and/or others) suggesting to support more complex options there. Again, this is just my opinion and as long as the keywords still exist, I would likely use them instead (you are still supporting `class` and `id` keywords right?).

> Since these classes are instantiated as one for every block, it doesn't quite make sense to pass them this way.

Ah right. So the config needs to be passed into the `Block` class directly, not the block-type class. I missed that detail. I think either (or both) of your suggested solutions would suffice. I prefer the first, but you would need to support both for those text-based config file situations (like MkDocs).

> ```python
> md = markdown.Markdown(
>     extensions=[Blocks(blocks={Admonition: {'types': ['note', 'warning', 'some-custom-type']}})]
> )
> ```
> 
> ```python
> md = markdown.Markdown(
>     extensions=[Blocks(blocks={'pymdownx.blocks.admonition:Admonition': {'types': ['note', 'warning', 'some-custom-type']}})]
> )
> ```

Of course, the following example which does use `extension_configs` would be functionally identical to both of the above examples:

```python
md = markdown.Markdown(
    extensions=['pymdownx.blocks'],
    extension_configs={
        'pymdownx.blocks': {
            'blocks': [{'pymdownx.blocks.admonition:Admonition': {'types': ['note', 'warning', 'some-custom-type']}}],
        }
    }
)
```

And the MkDocs equivalent would look like this:

```yaml
  - pymdownx.blocks:
      blocks:
        - pymdownx.blocks.admonition:Admonition:
          types:
            - note
            - warning
            - some-custom-type
```

**Reply by facelessuser** [reply]
> Let's focus on the `$` syntax first

I think there is some confusion about `$`. This is for all attributes, not classes specifically, that's why it isn't `.`. I guess we could change it to `attrs` or even revert it back to `attributes`. The idea was to provide a reserved word that is quick to type for a very common application (applying arbitrary attributes). The idea was that `$` would be "special" and allow people to use any keywords in their plugin without conflict. So we wouldn't be taking up an actual keyword name.

```
/// admonition | Title
    $: .class#id
```

Now, is using `$` a hill I'm willing to die on? Maybe not. I am interested in hearing other opinions on this. Before I make a decision.
Worst case we rename it to `attrs`.

Now as far as using a dictionary vs inline selectors to create the attributes. I am open to using dictionaries if people just outright hate inline attributes for `$` or `attrs`.  I'm not sure I completely understand the dislike of a compact approach, but I am far more interested in defending the HTML block inline approach than I am the general-purpose YAML attribute options.

If people want, we can change it to:

```
/// admonition | Title
    $: {class: name, id: some-id}

... or...

/// admonition | Title
    attrs: {class: name, id: some-id}
```

> Regarding removing the `type` keyword from admonitions, I understand the motivation. However, I wouldn't remove the keyword.

I'm fine with this.

> My opinion has always been that the option on the first line (after the `|`) should only ever be a single word.

What about multi-word titles? Maybe you meant to say a single option, not "word"? Regardless, I simply disagree with this. I can understand the argument to state it should be only one option or even the intent is to only use it for things like titles, but a plugin developer is going (ab)use it to do whatever they want anyway. I mentioned early on that I have no intention to place arbitrary restrictions on this, and if I did, I could not enforce it as it is trivial to workaround anything I put in place to limit this.

IMHO, the HTML block input is one input. I think it is a different kind of block than an Admontion and titles don't apply, but the point of these blocks is to make mine and eveyone's lives easier. I personally find the inline styling far more convenient. This is something I don't plan on changing, but since attributes can still be defined via YAML option, there's probably not much use debating this further.

> Again, this is just my opinion and as long as the keywords still exist, I would likely use them instead (you are still supporting `class` and `id` keywords right?).

Yes, you can still use attributes via the YAML config.  As discussed above, I guess the format of this is in contention, but I don't forsee issues resolving this.

> Ah right. So the config needs to be passed into the `Block` class directly, not the block-type class. I missed that detail. I think either (or both) of your suggested solutions would suffice. I prefer the first, but you would need to support both for those text-based config file situations (like MkDocs).

Yeah, I already have the code in place to accept the actual plugin object or the string, so that would not change. I don't mind simplifying the interface to just accept a single `block` input. If the plugin has no options, you'd have to specify an empty `dict` or maybe we could also accept `None`.

**Reply by waylan** [reply]
> I think there is some confusion about `$`. This is for all attributes, not classes specifically, that's why it isn't `.`. I guess we could change it to `attrs` or even revert it back to `attributes`.

That is my point. `$` has no discernable meaning. `attrs` (or `attributes`) does. Therefore others will have similar confusion as I had. IMO the confusion `$` is to cause for users is not worth the  savings in key strokes. 

Regarding the option after the `|`, we have disagreed on this from the beginning. We will have to agree to disagree here. And as the individual items can still be defined in the YAML config, there is no value in arguing about it. As an aside, I will admit that I was wrong to say it should be "one word" (my position has always been that it should be a single simple value).

> If the plugin has no options, you'd have to specify an empty `dict` or maybe we could also accept `None`.

I would expect it to default to `None` if no value is provided.  Of course, the tricky part is parsing the config. How do you handle `blocks={Admonition: {}}`  versus `blocks=[Admonition]`? While that later would be preferable for the user, the code is simpler if you only accept the former. Of then you also can get this: `blocks={Admonition: {}, OtherBlock: {key:value}}`. As the other block does get config values, the user needs to pass an empty `dict` into the first block. The more I think about this, the less I like the whole thing.

**Reply by facelessuser** [reply]
Okay, so here are the action items:

- Eliminate the need for `block_configs`
- Add back `type` option to Details and Admonitions.
- Rename `$` to `attrs`
- Revert `attrs` back to using key/value pairs.

I've probably spent more time on this plugin than I intended. I'm ready to roll out the Beta.
HTML blocks will remain as is. If the inline attributes are undesired, they can be avoided already.

I'll release the beta after these changes get in. Hopefully can have it all done by this evening.

**Reply by facelessuser** [reply]
> I would expect it to default to `None` if no value is provided. Of course, the tricky part is parsing the config. How do you handle `blocks={Admonition: {}}` versus `blocks=[Admonition]`? While that later would be preferable for the user, the code is simpler if you only accept the former. Of then you also can get this: `blocks={Admonition: {}, OtherBlock: {key:value}}`. As the other block does get config values, the user needs to pass an empty `dict` into the first block. The more I think about this, the less I like the whole thing.

It will be simply: `blocks={Plugin: None}` or `blocks={Plugin: {}}`. It kind of is what it is. It's a plugin system in a plugin system. It was weird from the start, but kind of necessary for what we are doing.

**Reply by facelessuser** [reply]
I am open to suggestions for other ways to handle the plugin and config. I'm probably not going to let it hold me up at this point though. I like `blocks` and `blocks_configs` being combined better than apart. It isn't perfect, but it is the best we have so far (I think at least).

**Reply by waylan** [reply]
I think in an ideal world, each block type would be its own Markdown extension rather than a plugin to a block extension. But I understand that that would be difficult to do as we only want the general block parsing once for all types. Although perhaps the various types could each be their own Markdown extensions which all have a dependency that requires that the general `block` extension be enabled as well. But then we could get into this weirdness where the order of the individual extensions matters when initiating them. It's probably not practical, but I thought I would mention it.

**Reply by facelessuser** [reply]
Yeah, I thought about this, but the order doesn't really matter as they all have unique block names. That's a requirement. The only thing that cares about order is the main blocks plugin.

You could maybe do something weird where you register a block, but what actually happens is that they check if the block manager is loaded, and if not, they register it and then register themselves within the manager.  This requires the developer to copy and paste this logic if they want to define their own block plugin. What we have now is much easier on the developer, and only slightly awkward for the user.  I'm not sure if it is worth the effort to try and restructure this.

**Reply by waylan** [reply]
> what actually happens is that they check if the block manager is loaded, and if not, they register it and then register themselves within the manager. This requires the developer to copy and paste this logic if they want to define their own block plugin.

Actually, a base class could implement this so that any subclass would get this behavior for free.

>  What we have now is much easier on the developer, and only slightly awkward for the user. I'm not sure if it is worth the effort to try and restructure this.

I have found that the best approach is usually to give preference to the user in situations like this. That would suggest that the effort would be worth it to the users. My concern is that it is currently overly complex to get anything working at all. Refactoring so that each block type was its own extension which appears to stand alone would greatly simplify this for the user.

That said, I am viewing this from the perspective of someone who has only read the documentation. I still have not looked at your code at all and am not clear on how much work it would be to do this refactor. I also understand that this has gone on for some time with multiple refactors and you are understandably reluctant to do any more.

**Reply by facelessuser** [reply]
I can see how complicated it might be. If it isn't too bad, I might give it a go. I can understand the desire to a more simple interface.

**Reply by facelessuser** [reply]
I think we may be able to subclass the extension class that handles the registering of processors. If we did that, we can probably drop the on_register event...I feel like there may be some complexity with reset events...I guess we'll see.

**Reply by facelessuser** [reply]
Okay, I got block extensions registering on their own. Moving forward, if you want to you'd like to create one these block extensions, you'll have to use the `pymdownx.blocks.BlockExtension` instead of `markdown.Extension`.

Instead of overriding `extendMarkdown`, you override `extendMarkdownBlocks`. That method receives `md` and `blocks`. `blocks` is used to register block plugins while `md` can be used to register normal extensions. We handle the abstraction in `extendMarkdown` which should not be overridden directly. Everything else is like usual.

```python
class AdmonitionExtension(BlocksExtension):
    """Admonition Blocks Extension."""

    def __init__(self, *args, **kwargs):
        """Initialize."""

        self.config = {
            "types": [
                ['note', 'attention', 'caution', 'danger', 'error', 'tip', 'hint', 'warning'],
                "Generate Admonition block extensions for the given types."
            ]
        }

        super().__init__(*args, **kwargs)

    def extendMarkdownBlocks(self, md, blocks):
        """Extend Markdown blocks."""

        blocks.register(Admonition, self.getConfigs())
```

I think this is definitely for the better. Now I have to update all the tests and update the docs 😢.

**Reply by facelessuser** [reply]
Everything is working, but we'll have to delay the beta until the end of the week as I don't feel like getting it all polished up today. But I think we are getting close to the finish line.

The plan is to release the beta. People can try it out and report any bugs. Feedback is still welcome, but I will be more critical of big changes moving forward. Then we'll go live in a non-beta release adding a disclaimer that these are "experimental". Assuming they are generally well received, we can remove the disclaimer at some point.

**Comment by facelessuser** [comment]
Ended up getting the beta out. The new discussion is here: https://github.com/facelessuser/pymdown-extensions/discussions/1961.

Now's the time to get your feedback in before this goes public.

Locking this discussion in favor of the Beta 1 discussion.
